cub/test/catch2_test_device_for_each_in_extents.cu

```diff
@@ -0,0 +1,190 @@
+static void fill_linear_impl(c2h::host_vector<T>& vector, const ExtentType&, size_t& pos, IndicesType... indices)
+{
+  vector[pos++] = {indices...};
+}
+
+template <int Rank = 0,
+          typename T,
+          typename ExtentType,
+          _CUB_TEMPLATE_REQUIRES(Rank < ExtentType::rank()),
+          typename... IndicesType>
+static void fill_linear_impl(c2h::host_vector<T>& vector, const ExtentType& ext, size_t& pos, IndicesType... indices)
+{
+  using IndexType = typename ExtentType::index_type;
+  for (IndexType i = 0; i < ext.extent(Rank); ++i)
+  {
+    fill_linear_impl<Rank + 1>(vector, ext, pos, indices..., i);
+  }
+}
+
+template <typename T, typename IndexType, size_t... Extents>
+static void fill_linear(c2h::host_vector<T>& vector, const cuda::std::extents<IndexType, Extents...>& ext)
+{
+  size_t pos = 0;
+  fill_linear_impl(vector, ext, pos);
+}
+
+/***********************************************************************************************************************
+ * Function Objects
+ **********************************************************************************************************************/
+
+template <typename IndexType, int Size>
+struct LinearStore
+{
+  using data_t = cuda::std::array<IndexType, Size>;
+
+  cuda::std::span<data_t> d_output_raw;
+
+  template <typename... TArgs>
+  __device__ void operator()(IndexType idx, TArgs... args)
+  {
+    static_assert(sizeof...(TArgs) == Size, "wrong number of arguments");
+    d_output_raw[idx] = {args...};
+  }
+};
+
+/***********************************************************************************************************************
+ * TEST CASES
+ **********************************************************************************************************************/
+
+using index_types =
+  c2h::type_list<int8_t,
+                 uint8_t,
+                 int16_t,
+                 uint16_t,
+                 int32_t,
+                 uint32_t
+#  if CUB_IS_INT128_ENABLED
+                 ,
+                 int64_t,
+                 uint64_t
+#  endif
+                 >;
+
+// int8_t/uint8_t are not enabled because they easily overflow
+using index_types_dynamic =
+  c2h::type_list<int16_t,
+                 uint16_t,
+                 int32_t,
+                 uint32_t
+#  if CUB_IS_INT128_ENABLED
+                 ,
+                 int64_t,
+                 uint64_t
+#  endif
+                 >;
+
+using dimensions =
+  c2h::type_list<cuda::std::index_sequence<>,
+                 cuda::std::index_sequence<5>,
+                 cuda::std::index_sequence<5, 3>,
+                 cuda::std::index_sequence<5, 3, 4>,
+                 cuda::std::index_sequence<3, 2, 5, 4>>;
+
+template <typename IndexType, typename Dimensions>
+struct build_extents;
+
+template <typename IndexType, size_t... Dimensions>
+struct build_extents<IndexType, cuda::std::index_sequence<Dimensions...>>
+{
+  using type = cuda::std::extents<IndexType, Dimensions...>;
+};
```

Comment:
Remark: functions are sometimes easier if you want to pattern patch a templated entity:
```suggestion
template <typename IndexType, typename Dimensions>
auto build_extents(IndexType, cuda::std::index_sequence<Dimensions...>) -> cuda::std::extents<IndexType, Dimensions...> {
  return {};
}
```
And then later `decltype(build_extents(index_type{}, dims{}))`.